src/shared/sgxlkl_enclave_config.c

```diff
@@ -0,0 +1,560 @@
+
+#define JU64(P, D) JPATH(P, return decode_uint64_t(parser, type, un, P, &D));
+#define JU32(P, D) JPATH(P, return decode_uint32_t(parser, type, un, P, &D));
+
+#define ALLOC_ARRAY(N, A, T)                              \
+    do                                                    \
+    {                                                     \
+        data->config->N = un->integer;                    \
+        data->config->A = calloc(un->integer, sizeof(T)); \
+        CHECKMEM(data->config->A);                        \
+    } while (0)
+
+static json_result_t json_read_callback(
+    json_parser_t* parser,
+    json_reason_t reason,
+    json_type_t type,
+    const json_union_t* un,
+    void* callback_data)
+{
+    static char* last_path = NULL;
+    json_callback_data_t* data = (json_callback_data_t*)callback_data;
+    size_t i = parser->path[parser->depth - 1].index;
+    sgxlkl_enclave_config_t* cfg = data->config;
+
+    switch (reason)
+    {
+        case JSON_REASON_NONE:
+            FAIL("unreachable");
+        case JSON_REASON_NAME:
+            break;
+        case JSON_REASON_BEGIN_OBJECT:
+            if (data->enforce_format)
+                last_path = make_path(parser);
+            break;
+        case JSON_REASON_END_OBJECT:
+            break;
+        case JSON_REASON_BEGIN_ARRAY:
+            if (MATCH("args"))
+                ALLOC_ARRAY(num_args, args, char*);
+            else if (MATCH("env"))
+                ALLOC_ARRAY(num_env, env, char*);
+            else if (MATCH("auxv"))
+                ALLOC_ARRAY(num_auxv, auxv, Elf64_auxv_t);
+            else if (MATCH("mounts"))
+                ALLOC_ARRAY(num_mounts, mounts, sgxlkl_enclave_mount_config_t);
+            else if (MATCH("host_import_env"))
+                ALLOC_ARRAY(num_host_import_env, host_import_env, char*);
+            else if (MATCH("wg.peers"))
+                ALLOC_ARRAY(
+                    wg.num_peers, wg.peers, sgxlkl_enclave_wg_peer_config_t);
+            else if (!(MATCH("clock_res")))
+                FAIL("unknown json array '%s'\n", make_path(parser));
+            break;
+        case JSON_REASON_END_ARRAY:
+            if (data->enforce_format)
+                last_path = make_path(parser);
+            break;
+        case JSON_REASON_VALUE:
+        {
+            if (data->enforce_format)
+            {
+                char* cur_path = make_path(parser);
+                if (last_path)
+                {
+                    if (strcmp(last_path, cur_path) > 0)
+                        FAIL("unsorted json: %s >= %s\n", last_path, cur_path);
+                    free(last_path);
+                }
+                last_path = cur_path;
+            }
+
+            JPATHT("format_version", JSON_TYPE_STRING, {
+                uint64_t format_version =
+                    _strtoul(un->string, NULL, 10, !data->enforce_format);
+                if (format_version < SGXLKL_ENCLAVE_CONFIG_VERSION)
```

Comment:
This is specifically craft so that the launcher can use a newer version with additional settings. I think you're thinking of a different kind of version number. This is not intended to check whether we switched from json to xml, but kept encoding the same config settings, instead, it should change when we add/remove settings, so more like the "schema version".

Reply:
I mean the schema version. Are saying it should be allowed for a customer to define a version 2 config and run it in a version 1 config SGX-LKL and any unknown fields are simply ignored? That sounds quite bad.

Reply:
No, I'm saying we reject everything but exact version matches. You are the one who's been saying we should ignore new settings in old images. So, I take it then that you want to ignore extra settings, but only if the version numbers match?

Reply:
Can you remove the format_version from this PR and open an issue to discuss this? It's not something we should solve here.